### ISSUE TITLE
Extended Query: Support Batch Execution

### ISSUE THREAD

**Issue by mpareja** [brianc/node-postgres#2257]
Hi!

The [Extended Query](https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY) protocol enables drivers to submit multiple BIND messages before SYNC. One of the big benefits of using Extended Queries is that you can efficiently send a batch of executions without waiting for round trips for each execution. Pairing that with prepared statements and some simplifications: you send a single PARSE, a bunch of BIND/EXECUTE and a SYNC to find out how things went.

In other words, you'd be able to support something like the following without needing 4 entire round trips. (I'm not _recommending_ this API since it would be a terrible breaking change.)

```javascript
await client.query({
  name: 'my_query',
  text: 'insert sometable (id, val) values ($1, $2)'
  values: [
    [ 1, 'asdf' ],
    [ 2, 'fdsa' ],
    [ 3, 'qwer' ],
    [ 4, 'uytr' ]
  ]
})
```

For more information, check out how the JDBC Postgres driver handles a batched execution. There are a few layers to dig through, but [this appears to be the core of the code that sends a batch of messages](https://github.com/pgjdbc/pgjdbc/blob/b80ebcef33f9c3801f01e28994746bc36abb2d21/pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java#L490-L509) and subsequently sends a single SYNC. NOTE: their driver imposes a limit of 128 records per batch as ([apparently](https://github.com/pgjdbc/pgjdbc/blob/b80ebcef33f9c3801f01e28994746bc36abb2d21/pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java#L1604-L1608)) further batching does not improve performance.

**Comment by rightaway** [comment]
Isn't it already possible in this case? https://github.com/brianc/node-postgres/issues/1190#issuecomment-619557934

**Comment by mpareja** [comment]
While the two approaches look similar, there are very different performance characteristics. Sending a large number of SQL statements with different parameters will perform much worse than sending a single prepared statement and binding many parameters to it.

**Comment by brianc** [comment]
yeah a proper "batched query" would be nice.  Probably a separate object you pass to `client.query` or `pool.query` that was like

```
const batch = new BatchQuery({
  name: 'optional',
  text: 'INSERT INTO foo (bar) VALUES ($1)',
  values: [
    ['first'],
    ['second']
  ]
})
const result = client.query(batch)
for (const res of result) {
  for (const row of res) {
  }
}
```

Then the batch query execution could throw if this is false for some validation:

```
for (const row of config.values) {
  if (!Array.isArray(config.values)) {
    throw new Error('Batch commands require each set of values to be an array. e.g. values: any[][]')
  }
}
```

something like that.  Then it would be explicit.

**Comment by damianobarbati** [comment]
@brianc is this supported?

**Comment by brianc** [comment]
> is this supported?

from a protocol perspective, yes. But I haven't actually implemented the code yet.

**Comment by iamkhush** [comment]
Hi @brianc , I am interesting contributing since I believe this would be helpful for my usecase in production.
Any guidance would be appreciated.

**Comment by brianc** [comment]
https://github.com/brianc/node-postgres/blob/master/packages/pg-cursor/index.js#L42

This is a good point of reference.

**Comment by aleclarson** [comment]
It would be great if this feature was implemented with support for multiple queries, rather than multiple value arrays only.

```ts
const batch = new BatchQuery({
  name: 'optional',
  queries: [
    ['INSERT INTO foo (bar) VALUES ($1)', ['first']],
    ['DELETE FROM foo WHERE id = $1', [1]],
  ],
})
```

**Comment by charmander** [comment]
@aleclarson That would be pipelining, not multiple bind.

**Comment by abenhamdine** [comment]
for pipelining, see the experiment in https://github.com/brianc/node-postgres/pull/2706

**Comment by iamkhush** [comment]
Hey @brianc @mpareja I have put up a PR.
Can you please have a look at the PR and provide feedback

Remarkably, the bench.ts file shows that inserts are getting 100% increase in query per second.
